File: src/mcp-server/core/jsonl_parser.py
"""
JSONL parser for Claude Code session files.

Parses session files in JSONL format and extracts messages and tool calls.
"""

import json
from pathlib import Path
from typing import List, Dict, Any, Optional
from dataclasses import dataclass


@dataclass
class Message:
    """Represents a message in the session."""
    role: str  # 'user', 'assistant', 'system'
    content: str
    timestamp: Optional[str] = None
    line_number: int = 0


@dataclass
class ToolCall:
    """Represents a tool call in the session."""
    tool_name: str
    input: Dict[str, Any]
    timestamp: Optional[str] = None
    line_number: int = 0


@dataclass
class SessionData:
    """Parsed session data."""
    session_id: str
    messages: List[Message]
    tool_calls: List[ToolCall]
    metadata: Dict[str, Any]


class JSONLParser:
    """Parses Claude Code session JSONL files."""

    def parse_file(self, file_path: Path) -> SessionData:
        """
        Parse a JSONL session file.

        Args:
            file_path: Path to JSONL file

        Returns:
            SessionData object

        Raises:
            FileNotFoundError: If file doesn't exist
            ValueError: If file is empty or invalid
        """
        if not file_path.exists():
            raise FileNotFoundError(f"Session file not found: {file_path}")

        messages = []
        tool_calls = []
        metadata = {}

        # Extract session ID from filename
        session_id = file_path.stem

        with open(file_path, 'r', encoding='utf-8') as f:
            for line_num, line in enumerate(f, 1):
                line = line.strip()
                if not line:
                    continue

                try:
                    data = json.loads(line)
                    self._process_line(data, line_num, messages, tool_calls, metadata)
                except json.JSONDecodeError as e:
                    # Skip malformed lines but log warning
                    print(f"Warning: Skipping malformed JSON at line {line_num}: {e}")
                    continue
                except Exception as e:
                    print(f"Warning: Error processing line {line_num}: {e}")
                    continue

        if not messages and not tool_calls:
            raise ValueError(f"No valid data found in session file: {file_path}")

        return SessionData(
            session_id=session_id,
            messages=messages,
            tool_calls=tool_calls,
            metadata=metadata
        )

    def _process_line(
        self,
        data: Dict[str, Any],
        line_num: int,
        messages: List[Message],
        tool_calls: List[ToolCall],
        metadata: Dict[str, Any]
    ) -> None:
        """Process a single JSONL line."""

        # Extract timestamp if available
        timestamp = data.get('timestamp')

        # Check if it's a message
        if 'role' in data:
            role = data['role']
            content = data.get('content', '')

            # Handle content that might be a list or dict
            if isinstance(content, list):
                # Extract text from content blocks
                text_parts = []
                for block in content:
                    if isinstance(block, dict):
                        if block.get('type') == 'text':
                            text_parts.append(block.get('text', ''))
                        elif 'text' in block:
                            text_parts.append(block['text'])
                content = '\n'.join(text_parts)
            elif isinstance(content, dict):
                content = content.get('text', str(content))
            elif not isinstance(content, str):
                content = str(content)

            if content:  # Only add non-empty messages
                messages.append(Message(
                    role=role,
                    content=content,
                    timestamp=timestamp,
                    line_number=line_num
                ))

        # Check if it's a tool call
        elif data.get('type') == 'tool_use':
            tool_name = data.get('name', '')
            tool_input = data.get('input', {})
            if tool_name:
                tool_calls.append(ToolCall(
                    tool_name=tool_name,
                    input=tool_input,
                    timestamp=timestamp,
                    line_number=line_num
                ))

        # Extract metadata
        if 'metadata' in data:
            metadata.update(data['metadata'])

    def get_user_messages(self, session_data: SessionData) -> List[Message]:
        """
        Get only user messages from session.

        Args:
            session_data: Parsed session data

        Returns:
            List of user messages
        """
        return [m for m in session_data.messages if m.role == 'user']

    def get_assistant_messages(self, session_data: SessionData) -> List[Message]:
        """
        Get only assistant messages from session.

        Args:
            session_data: Parsed session data

        Returns:
            List of assistant messages
        """
        return [m for m in session_data.messages if m.role == 'assistant']

    def get_tool_calls_by_name(self, session_data: SessionData, tool_name: str) -> List[ToolCall]:
        """
        Get tool calls filtered by tool name.

        Args:
            session_data: Parsed session data
            tool_name: Name of tool to filter by

        Returns:
            List of matching tool calls
        """
        return [tc for tc in session_data.tool_calls if tc.tool_name == tool_name]

    def extract_file_paths(self, session_data: SessionData) -> List[str]:
        """
        Extract file paths from tool calls.

        Args:
            session_data: Parsed session data

        Returns:
            List of unique file paths
        """
        file_paths = set()

        for tool_call in session_data.tool_calls:
            # Check common parameter names for file paths
            for param in ['file_path', 'path', 'file', 'filepath']:
                if param in tool_call.input:
                    path = tool_call.input[param]
                    if isinstance(path, str):
                        file_paths.add(path)

        return sorted(file_paths)

File: src/mcp-server-ts/src/core/jsonl-parser.ts
/**
 * JSONL parser for Claude Code session files.
 *
 * Parses session files in JSONL format and extracts messages and tool calls.
 */

import { readFileSync } from "fs";
import { logger } from "./logger.js";

export interface Message {
  role: "user" | "assistant" | "system";
  content: string;
  timestamp?: string;
  line_number: number;
}

export interface ToolCall {
  tool_name: string;
  input: Record<string, any>;
  timestamp?: string;
  line_number: number;
}

export interface SessionData {
  session_id: string;
  messages: Message[];
  tool_calls: ToolCall[];
  metadata: Record<string, any>;
  project_path?: string;  // Extracted from cwd in session file
}

export class JSONLParser {
  parseFile(filePath: string): SessionData {
    const content = readFileSync(filePath, "utf-8");
    const lines = content.split("\n");

    const messages: Message[] = [];
    const toolCalls: ToolCall[] = [];
    const metadata: Record<string, any> = {};
    let projectPath: string | undefined;

    // Extract session ID from filename
    const sessionId = filePath.split("/").pop()?.replace(".jsonl", "") || "unknown";

    for (let lineNum = 0; lineNum < lines.length; lineNum++) {
      const line = lines[lineNum].trim();
      if (!line) continue;

      try {
        const data = JSON.parse(line);

        // Extract project path from cwd field (appears in user/assistant messages)
        if (data.cwd && !projectPath) {
          projectPath = data.cwd;
        }

        this.processLine(data, lineNum + 1, messages, toolCalls, metadata);
      } catch (error) {
        logger.warn("jsonl-parser", `Warning: Skipping malformed JSON at line ${lineNum + 1}`, { error: error instanceof Error ? error.message : String(error) });
        continue;
      }
    }

    // Return empty session data for files with only metadata (empty sessions)
    // This is not an error - it's a valid case for newly created or unused sessions
    return {
      session_id: sessionId,
      messages,
      tool_calls: toolCalls,
      metadata,
      project_path: projectPath
    };
  }

  private processLine(
    data: Record<string, any>,
    lineNum: number,
    messages: Message[],
    toolCalls: ToolCall[],
    metadata: Record<string, any>
  ): void {
    // Handle Claude Code format: {type: "user"|"assistant", message: {role, content}}
    if ((data.type === "user" || data.type === "assistant") && data.message) {
      const msgData = data.message;
      const message: Message = {
        role: msgData.role || data.type,
        content: this.extractContent(msgData),
        timestamp: data.timestamp,
        line_number: lineNum
      };
      messages.push(message);
      return;
    }

    // Extract message (legacy format)
    if (data.type === "message" || data.role) {
      const message: Message = {
        role: data.role || "assistant",
        content: this.extractContent(data),
        timestamp: data.timestamp,
        line_number: lineNum
      };
      messages.push(message);
    }

    // Extract tool calls from tool_use content blocks in assistant messages
    if (data.message && data.message.content && Array.isArray(data.message.content)) {
      for (const block of data.message.content) {
        if (block.type === "tool_use") {
          const toolCall: ToolCall = {
            tool_name: block.name || "unknown",
            input: block.input || {},
            timestamp: data.timestamp,
            line_number: lineNum
          };
          toolCalls.push(toolCall);
        }
      }
    }

    // Extract tool calls (legacy format)
    if (data.type === "tool_use" || data.tool_name) {
      const toolCall: ToolCall = {
        tool_name: data.tool_name || data.name || "unknown",
        input: data.input || data.parameters || {},
        timestamp: data.timestamp,
        line_number: lineNum
      };
      toolCalls.push(toolCall);
    }

    // Extract metadata
    if (data.metadata) {
      Object.assign(metadata, data.metadata);
    }
  }

  /**
   * Sanitize content by removing Claude Code skill system metadata
   * This prevents local filesystem paths and system noise from polluting pattern detection
   */
  private sanitizeContent(content: string): string {
    let sanitized = content;

    // Remove skill system metadata patterns
    const systemPatterns = [
      // "Base directory for this skill: /path/to/skill\n\n"
      /^Base directory for this skill:.*?\n\n/s,
      // "<command-message>...</command-message>\n"
      /^<command-message>[\s\S]*?<\/command-message>\s*/,
      // "<command-name>...</command-name>\n"
      /^<command-name>[\s\S]*?<\/command-name>\s*/,
      // "<command-args>...</command-args>\n"
      /^<command-args>[\s\S]*?<\/command-args>\s*/,
    ];

    for (const pattern of systemPatterns) {
      sanitized = sanitized.replace(pattern, '');
    }

    return sanitized.trim();
  }

  private extractContent(data: Record<string, any>): string {
    let content = '';

    if (typeof data.content === "string") {
      content = data.content;
    } else if (Array.isArray(data.content)) {
      content = data.content
        .map((block: any) => {
          if (typeof block === "string") return block;
          if (block.type === "text") return block.text;
          return "";
        })
        .filter(Boolean)
        .join("\n");
    } else if (data.text) {
      content = data.text;
    }

    return this.sanitizeContent(content);
  }
}
